src/command/list/plain.rs

```diff
@@ -0,0 +1,498 @@
+                .and_then(|p| describe_package_managers(&[p.clone()])),
+            describe_packages(&packages),
+        ),
+        Toolchain::All {
+            runtimes,
+            package_managers,
+            packages,
+        } => (
+            describe_runtimes(&runtimes),
+            describe_package_managers(&package_managers),
+            describe_packages(&packages),
+        ),
+    };
+
+    match (runtimes, package_managers, packages) {
+        (Some(runtimes), Some(package_managers), Some(packages)) => {
+            Some(format!("{}\n{}\n{}", runtimes, package_managers, packages))
+        }
+        (Some(runtimes), Some(package_managers), None) => {
+            Some(format!("{}\n{}", runtimes, package_managers))
+        }
+        (Some(runtimes), None, Some(packages)) => Some(format!("{}\n{}", runtimes, packages)),
+        (Some(runtimes), None, None) => Some(format!("{}", runtimes)),
+        (None, Some(package_managers), Some(packages)) => {
+            Some(format!("{}\n{}", package_managers, packages))
+        }
+        (None, Some(package_managers), None) => Some(format!("{}", package_managers)),
+        (None, None, Some(packages)) => Some(format!("{}", packages)),
+        (None, None, None) => None,
+    }
+}
+
+fn describe_runtimes(runtimes: &[Node]) -> Option<String> {
+    if runtimes.is_empty() {
+        None
+    } else {
+        Some(
+            runtimes
+                .iter()
+                .map(|runtime| display_node(&runtime.source, &runtime.version))
+                .collect::<Vec<String>>()
+                .join("\n"),
+        )
+    }
+}
+
+fn describe_package_managers(package_managers: &[PackageManager]) -> Option<String> {
+    if package_managers.is_empty() {
+        None
+    } else {
+        Some(
+            package_managers
+                .iter()
+                .map(|package_manager| display_package_manager(&package_manager))
+                .collect::<Vec<String>>()
+                .join("\n"),
+        )
+    }
+}
+
+fn describe_packages(packages: &[Package]) -> Option<String> {
+    if packages.is_empty() {
+        None
+    } else {
+        Some(
+            packages
+                .iter()
+                .map(display_package)
+                .collect::<Vec<String>>()
+                .join("\n"),
+        )
+    }
+}
+
+fn describe_tool_set(name: &str, hosts: &[Package]) -> String {
+    hosts
+        .iter()
+        .filter_map(|package| display_tool(name, package))
+        .collect::<Vec<String>>()
+        .join("\n")
+}
+
+fn display_node(source: &Source, version: &Version) -> String {
+    format!("runtime {}{}", tool_version("node", version), source)
+}
+
+fn display_package_manager(package_manager: &PackageManager) -> String {
+    format!(
+        "package-manager {}{}",
+        tool_version(&package_manager.kind, &package_manager.version),
+        package_manager.source
+    )
+}
+
+fn package_source(package: &Package) -> String {
```

Comment:
This seems like it duplicates the existing `impl Display for Source`, would it be possible to leverage that so that we don't have formatting code that needs to change in multiple places? Thinking it through, it's kinda tricky so if there's not a simple solution maybe something to keep in mind for a follow-on issue.

Reply:
It does indeed, though I don't think there's any good way to eliminate the duplication at this point, because the types are distinct. 🤔 I've opened #527 to serve as an investigation issue around it, though.

Reply:
Yeah, I spent a few minutes trying to come up with something and there wasn't any obvious solution to prevent the duplication without introducing weird structures elsewhere, so I think the follow-on issue makes sense.